torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,193 @@
+            dist.broadcast_object_list(
+                object_list=object_list,
+                group=self.group,
+                src=self.coordinator_rank)
+        return object_list[0]
+
+    def gather_object(self, object: Any) -> Union[List[Any], None]:
+        if self.use_dist:
+            gather_objs = [None] * dist.get_world_size(self.group) if self.is_coordinator else None
+
+            dist.gather_object(
+                obj=object,
+                object_gather_list=gather_objs if self.is_coordinator else None,
+                dst=self.coordinator_rank,
+                group=self.group
+            )
+            result = gather_objs
+        else:
+            result = [object]
+        return result
+
+    def all_gather_object(self, object: Any) -> List[Any]:
+        if self.use_dist:
+            gather_objs = [None] * dist.get_world_size(self.group)
+
+            dist.all_gather_object(
+                object_list=gather_objs,
+                obj=object,
+                group=self.group
+            )
+        else:
+            gather_objs = [object]
+        return gather_objs
+
+    def scatter_object(self, object_list: List[Any]) -> Any:
+        if self.use_dist:
+            gather_result = [None]
+            dist.scatter_object_list(
+                scatter_object_output_list=gather_result,
+                scatter_object_input_list=object_list if self.is_coordinator else None,
+                src=self.coordinator_rank,
+                group=self.group
+            )
+
+            local_reply = gather_result[0]
+        else:
+            local_reply = object_list[0]
+        return local_reply
+
+    def reduce_scatter(self,
+        step: str,
+        map_fun: Callable[[], Any],
+        reduce_fun: Callable[[List[Any]], List[Any]]
+    ) -> Any:
+        try:
+            local_data = map_fun()
+        except BaseException as e:
+            local_data = e
+
+        all_data = self.gather_object(local_data)
+        all_results = None
+        if self.is_coordinator:
+            node_failures = {i: err for i, err in enumerate(all_data) if isinstance(err, BaseException)}
+
+            if len(node_failures) == 0:
+                try:
+                    all_results = reduce_fun(all_data)
+                except BaseException as e:
+                    node_failures[self.rank] = e
+
+            if len(node_failures) > 0:
+                all_results = [CheckpointException(step, node_failures)] * self.get_world_size()
+
+        result = self.scatter_object(all_results)
+        if isinstance(result, CheckpointException):
+            raise result
+        return result
+
+    def all_reduce(self,
+        step: str,
+        map_cb: Callable[[], Any],
+        reduce_cb: Callable[[List[Any]], Any]
+    ) -> Tuple[Any, Any]:
+        try:
+            local_data = map_cb()
+        except BaseException as e:
+            local_data = e
+
+        all_data = self.gather_object(local_data)
```

Comment:
hmmm all_reduce shouldn't use gather_object?

Reply:
same as the previous one. All `_object` variants just wrap c10d with support for dist disabled.